arbitrum-docs/for-devs/quickstart-solidity-hardhat.md

```diff
@@ -0,0 +1,302 @@
+
+
+### Review our dumb contract
+
+Here's our vending machine implemented as a Javascript class:
+
+```js title="VendingMachine.js"
+   class VendingMachine {
+        // state variables = internal memory of the vending machine
+        cupcakeBalances = {};
+        cupcakeDistributionTimes = {};
+
+        // Vend a cupcake to the caller
+        giveCupcakeTo(userId) {
+            if (this.cupcakeBalances[userId] === undefined) {
+                this.cupcakeBalances[userId] = 0;
+                this.cupcakeDistributionTimes[userId] = 0;
+            }
+
+            // Rule 1: The vending machine will distribute a cupcake to anyone who hasn't recently received one.
+            const fiveSeconds = 5000;
+            const userCanReceiveCupcake = this.cupcakeDistributionTimes[userId] + fiveSeconds <= Date.now();
+            if (userCanReceiveCupcake) {
+                this.cupcakeBalances[userId]++;
+                this.cupcakeDistributionTimes[userId] = Date.now();
+                console.log(`Enjoy your cupcake, ${userId}!`);
+                return true;
+            } else {
+                console.error("HTTP 429: Too Many Cupcakes (you must wait at least 5 seconds between cupcakes)");
+                return false;
+            }
+        }
+
+        getCupcakeBalanceFor(userId) {
+            return this.cupcakeBalances[userId];
+        }
+    }
+```
+
+The `VendingMachine` class is like a contract - it uses *state variables* and *code* to implement *predefined rules*. This implementation is useful because it automates cupcake distribution, but it's "dumb" because it's hosted by a centralized server that we (Offchain Labs) control. So it's technically possible for someone at Offchain to give their friends extra cupcakes by modifying the above contract's code.
+
+This is a problem because it means that you have to trust us, we have to trust our hosting provider. How do we keep our cupcakes secure without having to trust anyone?
+
+DApps solve this problem by using Ethereum smart contracts to build *trustless* end-user experiences that nobody in particular controls[^2]. This gives us a way to follow the second rule: *"The vending machine's rules can't be changed by anyone."* 
+
+When our `VendingMachine` contract is deployed to Ethereum, it'll be hosted by Ethereum's decentralized network of nodes. We won't be able to modify the contract's code after it's deployed[^3].
+
+Let's convert our dumb contract into a smart contract.
+
+
+### Configure your project directory
+
+Create a `decentralized-cupcakes` directory for your project. Navigate to this directory from VS Code's integrated terminal and install [hardhat](https://hardhat.org/hardhat-runner/docs/getting-started#overview):
+
+<!-- todo: toggles for npm / yarn? -->
```

Comment:
Given the instruction granularity earlier (i.e., installing IDE) might want to include yarn installation / directory set up, e.g.

- Install yarn https://classic.yarnpkg.com/lang/en/docs/install/#mac-stable
- make a new directory `mkdir my_cupcake_app`
- setup new yarn project `yarn init`